src/main/java/bio/terra/workspace/service/resource/controlled/flight/delete/DeleteControlledResourceGcsBucketFlight.java

```diff
@@ -0,0 +1,174 @@
+
+      FlightMap inputParameters = flightContext.getInputParameters();
+      UUID workspaceId =
+          UUID.fromString(inputParameters.get(WorkspaceFlightMapKeys.WORKSPACE_ID, String.class));
+      UUID resourceId =
+          UUID.fromString(
+              inputParameters.get(WorkspaceFlightMapKeys.ResourceKeys.RESOURCE_ID, String.class));
+
+      String projectId = workspaceService.getGcpProject(workspaceId);
+
+      WsmResource wsmResource = resourceDao.getResource(workspaceId, resourceId);
+      ControlledGcsBucketResource resource =
+          wsmResource.castControlledResource().castToGcsBucketResource();
+
+      final StorageCow storageCow = crlService.createStorageCow(projectId);
+
+      BucketCow bucket = storageCow.get(resource.getBucketName());
+
+      boolean bucketExists = true;
+      while (bucketExists) {
+        // We always replace the lifecycle rules. This covers the case where the step is rerun
+        // and covers the case where the rules are changed out of band of this operation.
+        BucketCow bucketCow =
+            bucket.toBuilder()
+                .setLifecycleRules(
+                    ImmutableList.of(
+                        new BucketInfo.LifecycleRule(
+                            BucketInfo.LifecycleRule.LifecycleAction.newDeleteAction(),
+                            BucketInfo.LifecycleRule.LifecycleCondition.newBuilder()
+                                .setAge(0)
+                                .build())))
+                .build();
+        bucketCow.update();
+        bucketExists = tryBucketDelete(bucket);
+        if (bucketExists) {
+          TimeUnit.HOURS.sleep(1);
+        }
+        deleteTries++;
+        if (deleteTries >= MAX_DELETE_TRIES) {
+          // This will cause the flight to fail.
+          throw new BucketDeleteTimeoutException(
+              String.format("Failed to delete bucket after %d hours", MAX_DELETE_TRIES));
+        }
+      }
+      return StepResult.getStepResultSuccess();
+    }
+
+    /**
+     * Try deleting the bucket. It will fail if there are objects still in the bucket.
+     *
+     * @param bucket bucket we should try to delete
+     * @return bucket existence: true if the bucket still exists; false if we deleted it
+     */
+    private boolean tryBucketDelete(BucketCow bucket) {
+      try {
+        logger.info("Attempting to delete bucket " + bucket.getBucketInfo().getName());
+        bucket.delete();
+        return false;
+      } catch (StorageException ex) {
+        logger.info(
+            "Attempt to delete bucket "
+                + bucket.getBucketInfo().getName()
+                + "failed, but that is OK we will try again",
+            ex);
+        return true;
+      }
+    }
+
+    @Override
+    public StepResult undoStep(FlightContext flightContext) throws InterruptedException {
+      return StepResult.getStepResultSuccess();
+    }
+  }
+
+  // TODO: this looks generic, so factor it out and share it
```

Comment:
+1, this feels like it could be a standalone `DeleteControlledResourceMetadataStep`